### PR TITLE
Kh dev authenticated user

### LINKED ISSUE #73: Story 7: Authenticated User
As a visitor
When I visit "/"
Then I should see a link for "Login"
And when I click "Login"
And I should be on the "/login" page
I should see a place to insert my credentials to login
And I should see a link to "Create Account"

As a visitor 
When I visit "/login
And when I click link "Create Account"
And I fill in my desired credentials
And I submit my information
Then my current page should be "/dashboard"
And I should see a message in the navbar that says "Logged in as SOME_USER"
And I should see my profile information
And I should not see a link for "Login"
And I should see a link for "Logout"

### PR DESCRIPTION
**Pull request by kheppenstall**
User authentication:
-Create login path and page
-Create new account page
-Create user model (with model tests and factory)
-Create user dashboard
-Add user logs in and out features (slightly infringing on story 8)
-Customizes navbar on `current_user`

Users need to be logged in from login page and after creating an account. So there is a `#login_user` in `ApplicationController`.

`current_user` helper method is use in navbar and user show view.

Story number 7.
Closes #73

Note: Does not include styling on views or authorization.

### CODE REVIEW

**Comment by kheppenstall** [comment]
@akintner @wlffann @tmikeschu Ready for a review. Its a big one. Would love feedback on use of `current_user` in views and how I'm using `#login` in `ApplicationController`.

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  context '#name' do
+    it 'concatenates first and last name' do
+      user = create(:user)
+
+      expect(user.name).to eq "#{user.first_name} #{user.last_name}"
+    end
+  end
+end
```
Comment (akintner) [inline spec/models/user_spec.rb]: I like the method check here, very nice!

**Comment by akintner** [comment]
I just looked through everything and it all looks great to me.

**Comment by wlffann** [comment]
I like the navbar customization!
All this looks really good, Kyle. Thanks for taking on this big task!
I'm good to merge.

File: app/views/shared/_nav.html.haml [hunk 2]
```diff
@@ -10,11 +10,18 @@
       %img{ "src" => "/scoopful-logo.svg", "alt" => "logo" }
   / Collect the nav links, forms, and other content for toggling
   .collapse.navbar-collapse.navbar-ex1-collapse
+    %ul.nav.navbar-nav.navbar-left
+      %li 
+        - if current_user
+          = link_to "Logged in as #{current_user.first_name}", dashboard_path
```
Comment (tmikeschu) [inline app/views/shared/_nav.html.haml]: Nice!

File: spec/features/sessions/user_logs_out_spec.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+      fill_in :email, with: user.email
+      fill_in :password, with: 'password'
+      click_on 'Enter'
+
+      click_on 'Logout'
+
+      expect(current_path).to eq root_path
+      expect(page).to have_link 'Login'
+      expect(page).to_not have_content "Logged in as #{user.first_name}"
```
Comment (tmikeschu) [inline spec/features/sessions/user_logs_out_spec.rb]: I like to check non expectations with expectations in the beginning, e.g., `expect(page).to have_content "logged in as..."` on line 13 or something

**Comment by tmikeschu** [comment]
Nice work, Kyle! Very comprehensive.

**Pull request merged**